Choosing actual layer table owner

juanignaciosl opened the pull request:
This fixes #8974

juanignaciosl commented:
What to test:

- Organization users:
  - Create map with public, own table. Check public view.
  - Add layer with public, own table. Check public view.
  - Create map with private, own table. Check public view.
  - Add layer with private, own table. Check public view.
  - Create map with public, shared table. Check public view.
  - Add layer with public, shared table. Check public view.
  - Create map with private, shared table. Check public view.
  - Add layer with private, shared table. Check public view.
- Normal user
  - Create map with public, own table. Check public view.
  - Add layer with public, own table. Check public view.
  - Create map with private, own table. Check public view.
  - Add layer with private, own table. Check public view.

... for Editor and Builder

😢

juanignaciosl commented:
@gfiorav  could you do a first CR, please? Disclaimer: I haven't 100% tested it yet (see the test cases comment and feel free to add anything you think), but this is slow and annoying to test and I want to discuss the approach with you before. This change is huge and risky, the more eyes and brains aware of this, the better.

When we get to a final fix I'll do a test suite covering this before merging, because this can't happen again U_U

cc @javitonino , as he will be asked for CR later.

gfiorav commented on lib/carto/named_maps/template.rb:
```diff
@@ -167,6 +167,19 @@ def visibility_wrapped_sql(sql, index)
+        # Inspired from CartoDB::LayerModule::Presenter#default_query_for
+        layer_options = layer.options
+        user_name = layer_options['user_name']
+        table_name = layer_options['table_name']
```
I'd symbolise fetch these

gfiorav commented on lib/carto/named_maps/template.rb:
```diff
@@ -167,6 +167,19 @@ def visibility_wrapped_sql(sql, index)
+          %{ select * from "#{user_name}".#{table_name} }
+        else
+          layer.wrapped_sql(user)
+        end
```
You're not taking into account the possibility that a `:query` and/or a `:query_wrapper` is provided in layer options and has `<= sql >` that needs to be replaced by the sql you calculate here. Look at `wrapped_sql`:

```ruby
def wrapped_sql(user)
  query = options[:query]

  sql = if query.present?
          query
        else
          "SELECT * FROM #{qualified_table_name(user)}"
        end

  query_wrapper = options[:query_wrapper]
  sql = query_wrapper.gsub('<%= sql %>', sql) if query_wrapper.present? && torque?

  sql
end
```

juanignaciosl replied:
Proposals?

juanignaciosl replied:
PS: your comment is collapsed by Github because of an unrelated commit.

gfiorav replied:
Maybe have `wrapper_sql` receive params so to overwrite the normal `"SELECT * FROM #{qualified_table_name(user)}"` behaviour or overwrite the `:query` with the right params. It's ugly, but `:query` will have bad data (other user schema) that needs to be sanitised in this case.

gfiorav commented:
I've made some comments regarding some concerns I have!

juanignaciosl commented:
@gfiorav I've taken into account your two comments and moved a symbolized version of the patch from the named map to layer `wrapped_sql` method, so I think that I fix your relevant catch. Take a look at the commit and the full code while I keep testing this, please. I've also added some code to avoid double _schematization_ of the table name.

PS: create new comment threads for the next insights, as Github won't expand older ones once they've been collapsed.

gfiorav commented on app/controllers/carto/api/layers_controller.rb:
```diff
@@ -15,8 +15,8 @@ def layers_by_map
-            Carto::Api::LayerPresenter.new(layer, { viewer_user: current_user, user: @parent.user }).to_poro
-          }
+          Carto::Api::LayerPresenter.new(layer, viewer_user: current_user, user: owner_user(layer)).to_poro
+        }
```
only if you need to make a new commit anyway: `do...end` for multiple lines is nice :)

juanignaciosl replied:
👍

gfiorav commented on app/models/carto/layer.rb:
```diff
@@ -140,12 +145,20 @@ def user
+                %{ select * from "#{user_name}".#{table_name} }
+              else
+                "SELECT * FROM #{qualified_table_name(user)}"
+              end
```
Yes, this feels a lot safer now!

gfiorav commented on app/models/carto/layer.rb:
```diff
@@ -140,12 +145,20 @@ def user
+              table_name = options[:table_name]
+
+              if table_name.present? && !table_name.include?('.') && user_name.present? && user.username != user_name
+                %{ select * from "#{user_name}".#{table_name} }
```
Cosmetic, of course, but why not use caps for SQL reserved words and use `%{}`? If there is a good reason for any of them, please update line 160 as well :)

juanignaciosl replied:
We have no style convention about SQL caps. In general I prefer small caps but I don't really care. If @javitonino  and you agree upon any criteria I'll follow it. I'll also use it for new and changed code.

I prefer not adding anything smart for extracting `select *`. It's actual duplication, but any alternative is more complex and doesn't add any value, IMHO.

gfiorav commented on app/models/carto/layer.rb:
```diff
@@ -177,7 +190,10 @@ def affected_table_names
       return[] if options.empty?
-      ::Table.get_all_user_tables_by_names([options.symbolize_keys[:table_name]], user)
+      user_name = options.symbolize_keys[:user_name]
+      table_name = options.symbolize_keys[:table_name]
```
If you symbolise here, you should also symbolise 154 and 155.

juanignaciosl replied:
Yep, and only once. On it!

gfiorav commented on app/models/layer.rb:
```diff
@@ -210,7 +214,10 @@ def tables_from_query_option
   def tables_from_table_name_option
-    ::Table.get_all_by_names([options.symbolize_keys[:table_name]], user)
+    user_name = options.symbolize_keys[:user_name]
+    table_name = options.symbolize_keys[:table_name]
```
same here

juanignaciosl replied:
👍

gfiorav commented:
It looks safe to me!

You could deploy this now. If you find any improvements to make, maybe take my other comments into account.

📦 !

juanignaciosl commented:
No, I won't 🚢  yet. Best scenario: tomorrow early morning. I still have to run the tests and, if possible, cover the changes.

juanignaciosl commented:
@javitonino could you do a second CR, please? Editor tests with normal layers have worked. I can't test Torque because of #9017, and I've just seen an error with the builder, probably unrelated.

javitonino commented on app/controllers/carto/api/layers_controller.rb:
```diff
@@ -66,6 +66,19 @@ def map_from_params
+      def owner_user(layer)
+        if current_user.nil? || @parent.user.id != current_user.id
+          # This keeps backwards compatibility with map user assignment. See #8974
+          @parent.user
```
Does this work in the following case?
 - New visualization, so `visualization.user == map.user != dataset.user`
 - Visitor is anonymous, so `current_user == nil`

I think in that case, it goes into this code path, and takes the user name from the map, which would be incorrect (it should take the user from the layer/dataset).

juanignaciosl replied:
The user name from the map (now) is correct, it's the same than layer's and visualization's.

javitonino replied:
I think I see how it works now. The passed username would be the owner of the visualization, not of the shared dataset. So, `viewer_user = nil`, `user = vis_user`.  In `Carto::LayerPresenter.qualify_table_name` I think it would go through the qualified table name branch and work.

A mess (not a comment of your PR) but it works. :sparkles:

juanignaciosl replied:
Yup, really messy. When the old editor is gone I'll gather coders with torches and burn a lot of code around this >_<

javitonino commented:
I came up with a case where this might be wrong, although I may be totally mistaken. I will take another look first thing tomorrow, with a fresh head.

juanignaciosl commented:
@javitonino hold on, because this won't work with the builder U_U.

juanignaciosl commented:
@javitonino although I still have to do deep testing, latest commit should fix the builder issue, you can begin the CR whenever you want.

javitonino commented on app/controllers/carto/api/vizjson3_presenter.rb:
```diff
@@ -558,14 +558,12 @@ def layer_name
 
-        return table_name unless layer_alias && !layer_alias.empty?
-        layer_alias
+        layer_alias && !layer_alias.empty? ? layer_alias : table_name
```
`layer_alias.present?` may be cleaner here

juanignaciosl replied:
Sure. I only refactored the double exit point without seeing beyond that.

javitonino commented on app/controllers/carto/api/vizjson3_presenter.rb:
```diff
@@ -558,14 +558,12 @@ def layer_name
-        query
+      def sql_from(layer)
+        query = layer.options.fetch('query', '')
+        query.nil? || query.empty? ? layer.default_query : query
```
Same here with `layer.blank?`. Mainly, I don't like trusting the order of operations when ternary operators are involved.

juanignaciosl replied:
`.present?` 👍

javitonino replied:
`.blank?` is the antonym of `.present?`. But `present?` and reversing the condition works too.

juanignaciosl replied:
Yes, I prefer using `present?` and reversing. `blank?` is better if there's only one condition (to avoid `!whatever.present?`), but if there's an `else` being positive first is easier to read.

javitonino commented on app/models/carto/layer.rb:
```diff
@@ -177,7 +197,11 @@ def affected_table_names
+      sym_options = options.symbolize_keys
+      user_name = sym_options[:user_name]
+      table_name = sym_options[:table_name]
+      schema_prefix = user_name.present? && table_name.present? && !table_name.include?('.') ? %{"#{user_name}".} : ''
```
I'm alergic to trusting the order of operations with ternary operators in the mix :stuck_out_tongue:

juanignaciosl replied:
😆  how come? I don't mind changing it to an if, but ternary operation condition evaluation order is no different than an if ;-)

javitonino replied:
PHP + an aweful codestyle with nested ternaries. Did you know that some languages have right-associative ternaries and others do left-associativity? Not something you wold have to know because you shouldn't use more than one ternary per-expression. Also, `and`/`or` operators usually have lower precedence than ternaries but `&&`/`||` have higher one.

Anyway, this code is fine, easy to read and works as expected. I may have a little bit of PTSD from previous projects.

juanignaciosl replied:
:D I'll keep it as a treatment for your PTSD then :-)

juanignaciosl commented on app/models/carto/layer.rb:
```diff
@@ -140,18 +146,32 @@ def user
-      sql = query_wrapper.gsub('<%= sql %>', sql) if query_wrapper.present? && torque?
+    def default_query(user = nil)
+      sym_options = options.symbolize_keys
+      query = options[:query]
```
outch, good catch!

javitonino commented:
Looks good. Was thinking about other potentially problematic pieces of code but this should cover everything I can think of right now. Good job in tracking all of this, I can only imagine... :+1:

juanignaciosl commented:
@javitonino , @gfiorav thank you for the CRs. I'll test and cover. I'll ping you again before the final merge. :fingers_crossed:

juanignaciosl commented:
Ok, @javitonino (back), @alonsogarciapablo (front) , last CR before merging?

javitonino commented:
:+1:

alonsogarciapablo commented on lib/assets/javascripts/cartodb3/data/layer-definitions-collection.js:
```diff
@@ -83,11 +83,10 @@ module.exports = Backbone.Collection.extend({
       var sourceId = nodeIds.next(o.letter);
 
       if (o.table_name && (!o.source || o.source === sourceId)) {
-        o.table_name = o.table_name && o.table_name.replace(/"/g, '');
```
Not sure why this was introduced in the first place... why are we adding the double quotes back? `SELECT foo FROM user.name` to `SELECT bar FROM "user".name`?

juanignaciosl replied:
In PostgreSQL double quoting schemas or table names is useful to avoid problems with special characters. For example, for `juanignaciosl-ded01` user you need to use double quotes because of the dash. I asked @xavijam about the replacement and he said that "about that, I removed them on purpose, I didn’t know they were necessary".

alonsogarciapablo replied:
Thanks for the explanation! 👍 👍 👍

juanignaciosl commented:
Thank you, @alonsogarciapablo 

Master merged (and some redundant parenthesis that I previously forgot pushing, @javitonino ). I'll merge and deploy after 🍏

juanignaciosl commented:
Had to do an extra merge because of package.json, but last test passed, so I'm deploying.
